src/main/scala/is/hail/expr/ir/Compile.scala

```diff
@@ -317,8 +317,105 @@ object Compile {
       case AggIn(_) | AggMap(_, _, _, _) | AggSum(_, _) =>
         throw new RuntimeException(s"Aggregations must be extracted with ExtractAggregators before compilation: $ir")
 
+      case MakeSet(args, elementType) =>
+        val srvb = new StagedRegionValueBuilder(fb, TArray(elementType))
+        val addElement = srvb.addIRIntermediate(elementType)
+        val mvargs = args.map(compile(_))
+        val sortedArray = Code(
+          srvb.start(args.length, init = true),
+          Code(mvargs.map { case (dov, m, v) =>
+            Code(dov, m.mux(srvb.setMissing(), addElement(v)), srvb.advance())
+          }: _*),
+          InplaceSort(region, fb, mb, srvb.offset, TArray(elementType)),
+          srvb.offset)
+
+        present(sortedArray)
+      case SetAdd(set, element, elementType) =>
+        val tArray = TArray(elementType)
+        val srvb = new StagedRegionValueBuilder(fb, tArray)
+        val (doset, mset, vset) = compile(set)
+        val (doelement, melement, velement) = compile(element)
+        val len = fb.newLocal[Int]
+        val i = fb.newLocal[Int]
+        val s = fb.newLocal[Long]
+        val mx = mb.newBit()
+        val x = fb.newLocal[Long]
+        val storedElement = fb.newLocal()(TypeToTypeInfo(tArray.elementType)).asInstanceOf[LocalRef[Any]]
+        val elementPointer = fb.newLocal[Long]
+        val ord = tArray.unsafeOrdering(true)
+        val sortedArray = Code(
+          doelement,
+          s := coerce[Long](vset),
+          len := TContainer.loadLength(region, s),
+          melement.mux(
+            Code(
+              srvb.start(len + 1, init = true),
+              i := 0,
+              Code.whileLoop(i < len,
+                srvb.addRegionValue(elementType)(tArray.loadElement(region, s, i)),
+                i++),
+              srvb.setMissing()),
+            Code(
+              storedElement := velement,
+              elementPointer := toPointer(region, tArray.elementType)(storedElement),
+              srvb.start(len + 1, init = true),
+              (len.ceq(0)).mux(
+                Code(
+                  melement.mux(
+                    srvb.setMissing(),
+                    srvb.addIRIntermediate(elementType)(velement))),
+                Code(
+                  i := 0,
+                  mx := tArray.isElementMissing(region, s, i),
+                  x := mx.mux(0L, tArray.loadElement(region, s, i)),
+                  Code.whileLoop(i < len && !mx && (ord.compare(region, x, region, elementPointer)(fb, mb) < 0),
+                    mx.mux(srvb.setMissing(), srvb.addRegionValue(elementType)(x)),
+                    mx := tArray.isElementMissing(region, s, i),
+                    x := mx.mux(0L, tArray.loadElement(region, s, i)),
+                    i++),
```

Comment:
I think the i++ needs to come before the two assignments.

Reply:
done.